libnavui-androidauto/src/test/java/com/mapbox/androidauto/car/feedback/ui/CarFeedbackIconDownloaderTest.kt

```diff
@@ -0,0 +1,197 @@
+        verify(exactly = 0) { screen.invalidate() }
+    }
+
+    @Test
+    fun `remote icon starts downloading if retrieved for the first time`() {
+        coroutineRule.runBlockingTest {
+            mockSuccessfulRequest(SHORT_DELAY)
+
+            val actualIcon = downloader.getOrDownload(remoteIcon)
+
+            assertNull(actualIcon)
+            verify(exactly = 1) { requestManager.load(uri) }
+            verify(exactly = 0) { screen.invalidate() }
+        }
+    }
+
+    @Test
+    fun `null is returned, while icon is being downloaded`() {
+        coroutineRule.runBlockingTest {
+            mockSuccessfulRequest(LONG_DELAY)
+
+            downloader.getOrDownload(remoteIcon)
+            advanceTimeBy(SHORT_DELAY)
+            val actualIcon = downloader.getOrDownload(remoteIcon)
+
+            assertNull(actualIcon)
+            verify(exactly = 1) { requestManager.load(uri) }
+            verify(exactly = 0) { screen.invalidate() }
+        }
+    }
+
+    @Test
+    fun `screen is invalidated, after icon is downloaded`() {
+        coroutineRule.runBlockingTest {
+            mockSuccessfulRequest(SHORT_DELAY)
+
+            downloader.getOrDownload(remoteIcon)
+            advanceTimeBy(LONG_DELAY)
+
+            verify(exactly = 1) { requestManager.load(uri) }
+            verify(exactly = 1) { screen.invalidate() }
+        }
+    }
+
+    @Test
+    fun `actual icon is returned, after icon is downloaded`() {
+        coroutineRule.runBlockingTest {
+            mockSuccessfulRequest(SHORT_DELAY)
+
+            downloader.getOrDownload(remoteIcon)
+            advanceTimeBy(LONG_DELAY)
+            val actualIcon = downloader.getOrDownload(remoteIcon)
+
+            val expectedIcon = IconCompat.createWithBitmap(downloadedBitmap)
+            assertEquals(CarIcon.Builder(expectedIcon).build(), actualIcon)
+            verify(exactly = 1) { requestManager.load(uri) }
+            verify(exactly = 1) { screen.invalidate() }
+        }
+    }
+
+    @Test
+    fun `error icon is returned, if download took too much time`() {
+        coroutineRule.runBlockingTest {
+            mockSuccessfulRequest(LONG_DELAY)
+
+            downloader.getOrDownload(remoteIcon)
+            advanceTimeBy(LONG_DELAY)
+            val actualIcon = downloader.getOrDownload(remoteIcon)
+
+            assertEquals(CarIcon.ERROR, actualIcon)
+            verify(exactly = 1) { requestManager.load(uri) }
+            verify(exactly = 1) { screen.invalidate() }
+        }
+    }
+
+    @Test
+    fun `error icon is returned, if download failed`() {
+        coroutineRule.runBlockingTest {
+            mockFailedRequest()
+
+            downloader.getOrDownload(remoteIcon)
+            advanceTimeBy(LONG_DELAY)
+            val actualIcon = downloader.getOrDownload(remoteIcon)
+
+            assertEquals(CarIcon.ERROR, actualIcon)
+            verify(exactly = 1) { requestManager.load(uri) }
+            verify(exactly = 1) { screen.invalidate() }
```

Comment:
Would be good to include a test that requests more than one

``` kotlin
val actualIcon1 = downloader.getOrDownload(remoteIcon1)
val actualIcon2 = downloader.getOrDownload(remoteIcon2)
val actualIcon3 = downloader.getOrDownload(remoteIcon3)
```

Small optimization but it seems it should only invalidate the screen one time